internal/mining/mining_harness_test.go

```diff
@@ -0,0 +1,1437 @@
+
+	// Add the stakebase input.
+	vote := wire.NewMsgTx()
+	stakebaseOutPoint := wire.NewOutPoint(&chainhash.Hash{}, ^uint32(0),
+		wire.TxTreeRegular)
+	stakebaseInput := wire.NewTxIn(stakebaseOutPoint, subsidy, nil)
+	vote.AddTxIn(stakebaseInput)
+
+	// Add the ticket input.
+	spendOut := txOutToSpendableOut(ticket, 0, wire.TxTreeStake)
+	ticketInput := wire.NewTxIn(&spendOut.outPoint, int64(spendOut.amount), nil)
+	ticketInput.BlockHeight = uint32(best.Height)
+	ticketInput.BlockIndex = 5
+	vote.AddTxIn(ticketInput)
+
+	// Add the block reference output.
+	blockRefScript, _ := txscript.GenerateSSGenBlockRef(best.Hash,
+		uint32(best.Height))
+	vote.AddTxOut(wire.NewTxOut(0, blockRefScript))
+
+	// Create the vote script.
+	voteBits := stake.VoteBits{Bits: uint16(0xff), ExtendedBits: []byte{}}
+	voteScript, err := newVoteScript(voteBits)
+	if err != nil {
+		return nil, err
+	}
+	vote.AddTxOut(wire.NewTxOut(0, voteScript))
+
+	// Create P2SH scripts for the ticket outputs.
+	for i, hash160 := range ticketHash160s {
+		scriptFn := txscript.PayToSSGenPKHDirect
+		if ticketPayKinds[i] { // P2SH
+			scriptFn = txscript.PayToSSGenSHDirect
+		}
+		// Error is checking for a nil hash160, just ignore it.
+		script, _ := scriptFn(hash160)
+		vote.AddTxOut(wire.NewTxOut(voteRewardValues[i], script))
+	}
+
+	// Perform any transaction munging just before signing.
+	for _, f := range mungers {
+		f(vote)
+	}
+
+	// Sign the input.
+	inputToSign := 1
+	redeemTicketScript := ticket.MsgTx().TxOut[0].PkScript
+	signedScript, err := txscript.SignTxOutput(m.chainParams, vote,
+		inputToSign, redeemTicketScript, txscript.SigHashAll, m, m,
+		vote.TxIn[inputToSign].SignatureScript, m.chain.isTreasuryAgendaActive)
+	if err != nil {
+		return nil, err
+	}
+
+	vote.TxIn[0].SignatureScript = m.chainParams.StakeBaseSigScript
+	vote.TxIn[1].SignatureScript = signedScript
+
+	return dcrutil.NewTx(vote), nil
+}
+
+// CountTotalSigOps returns the total number of signature operations for the
+// given transaction.
+func (m *miningHarness) CountTotalSigOps(tx *dcrutil.Tx) (int, error) {
+	txType := stake.DetermineTxType(tx.MsgTx(), m.chain.isTreasuryAgendaActive)
+	return m.txSource.CountTotalSigOps(tx, txType)
+}
+
+// AddTransactionToTxSource adds the given transaction to the tx source.
+func (m *miningHarness) AddTransactionToTxSource(tx *dcrutil.Tx) ([]*dcrutil.Tx, error) {
+	return m.txSource.ProcessTransaction(tx)
+}
+
+// RemoveTransactionFromTxSource removes the given transaction from the tx
+// source.
+func (m *miningHarness) RemoveTransactionFromTxSource(tx *dcrutil.Tx, removeRedeemers bool) {
+	m.txSource.RemoveTransaction(tx, removeRedeemers,
+		m.chain.isTreasuryAgendaActive)
+}
+
+// AddFakeUTXO creates a fake mined utxo for the provided transaction.
+func (m *miningHarness) AddFakeUTXO(tx *dcrutil.Tx, blockHeight int64, blockIndex uint32) {
+	m.chain.utxos.AddTxOuts(tx, blockHeight, blockIndex,
+		m.chain.isTreasuryAgendaActive)
```

Comment:
The idea was that the tests can set/unset `m.chain.isTreasuryAgendaActive`, `isHeaderCommitmentsAgendaActive`, etc. depending on what it is trying to simulate.

Reply:
Right, I understand that part of having the variable in the harness for the general case and it both makes sense and is a good idea I think, but I was really referring more to the fact that each individual utxo necessarily comes from some other place than the current tip of whatever the harness is associated with and therefore does not necessarily match whatever the current mocked state of the agenda is for the chain.

This would mean the caller would end up having to do something like the following for every utxo being mocked from some historical standpoint:

```Go
curTrsyActive := harness.chain.isTreasuryAgendaActive
harness.AddFakeUTXO(tx, someHeight, someIndex)
harness.chain.isTreasuryAgendaActive = curTrsyActive
```

Reply:
True, given that it is passing in the height and index as parameters, it probably makes more sense to pass isTreasuryAgendaActive as a parameter as well.  I'll update it to that.

Reply:
Updated this and also rebased on master.